crates/sui-core/src/authority/authority_notify_read.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::authority::AuthorityStore;
+use crate::checkpoints::causal_order_effects::EffectsStore;
+use async_trait::async_trait;
+use either::Either;
+use futures::future::join_all;
+use parking_lot::Mutex;
+use parking_lot::MutexGuard;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::HashMap;
+use std::future::Future;
+use std::hash::{Hash, Hasher};
+use std::mem;
+use std::pin::Pin;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use sui_types::base_types::TransactionDigest;
+use sui_types::error::SuiResult;
+use sui_types::messages::TransactionEffects;
+use tokio::sync::oneshot;
+
+#[allow(dead_code)]
+#[async_trait]
+pub trait EffectsNotifyRead {
+    /// This method reads transaction effects from database.
+    /// If effects are not available immediately, the method blocks until they are persisted
+    /// in the database.
+    ///
+    /// This method **does not** schedule transactions for execution - it is responsibility of the caller
+    /// to schedule transactions for execution before calling this method.
+    async fn notify_read(
+        &self,
+        digests: Vec<TransactionDigest>,
+    ) -> SuiResult<Vec<TransactionEffects>>;
+}
+
+type Registrations<V> = Vec<oneshot::Sender<V>>;
+
+pub(crate) struct NotifyRead<K, V> {
+    pending: Vec<Mutex<HashMap<K, Registrations<V>>>>,
+    count_pending: AtomicUsize,
+}
+
+impl<K: Eq + Hash + Clone, V: Clone> NotifyRead<K, V> {
+    pub fn new() -> Self {
+        let pending = (0..255).map(|_| Default::default()).collect();
+        let count_pending = Default::default();
+        Self {
+            pending,
+            count_pending,
+        }
+    }
+
+    /// Asynchronously notifies waiters and return number of remaining pending registration
+    pub fn notify(&self, key: &K, value: &V) -> usize {
+        let registrations = self.pending(key).remove(key);
+        let Some(registrations) = registrations else {
+            return self.count_pending.load(Ordering::Relaxed);
+        };
+        let rem = self
+            .count_pending
+            .fetch_sub(registrations.len(), Ordering::Relaxed);
+        for registration in registrations {
+            registration.send(value.clone()).ok();
+        }
+        rem
+    }
+
+    pub fn register_all(&self, keys: Vec<K>) -> Vec<Registration<K, V>> {
+        self.count_pending.fetch_add(keys.len(), Ordering::Relaxed);
```

Comment:
it seems that we can end up with the wrong pending count if we pass in duplicate keys. I'm not sure if this is a problem worth worrying about.

Reply:
I am actually not sure this is true. It won't be super optimal (we will have multiple registrations for same key), but I think as far as counter is concerned the value will be correct?